Add saving of prefix to WeMod dir to avoid rebuilding

marvin1099 opened the pull request:
Added a command to copy the game prefix to the wemod dir and back to make new games use the ready wine prefix, that will avoid rebuilding of the wine prefix
This is what I tried to asked for in [Issue 31](https://github.com/DaniAsh551/wemod-launcher/issues/31).
Would be nice if you could add it.

DaniAsh551 reviewed (changes requested):
Hey @marvin1099 , thank you for your contributions.

Actually, this idea has been a long standing request among users, but there is more to this than meets the eye.
For example, at the top of my head (I am just writing these without any plans - so please forgive me if I make a mistake), if we are to properly implement this feature, we need to address these concerns:
- Prefixes are not always compatible (atleast for the components we use in WeMod and this tool) between proton versions (a recent example is, with GE-Proton7-3X and GE-Proton7-4X).  One way to workaround this is to create a copy of the prefix for each Proton version, but this also comes with its own set of headaches, which include:

  * We have to track the games using this version of proton and seamlessly switch when the user selects a different proton version - doing this without loss of data is a bonus
   * When there is no game using this version of proton, we must delete the prefix from wemod-launcher, this once again - would require us to keep track of the versions the different games are using
    * This approach of setting up a prefix would increase the storage cost (at the rate of one extra prefix cost per each version of proton used) - but this isnt a major issue. One potential workaround for this approach would be symlinking the prefix or bind-mounting it. This would again introduce another challenge where if the user modifies one prefix, it would affect all the games using the same proton version. A potential fix for this would again be to use OverlayFS mounts, but this would need explicit support from the system and would require us to track the prefixes for the games in some manner
    
    In order to overcome these issues, we could diff the prefix before and after wemod installation and apply that diff onto any new prefixes, but this is essentially doing the same thing as we are, just with extra steps.
    
I, by no means mean to discourage you, infact, I love that you are actively trying to improve the situation. But I feel that this feature, at its current implementation, has the potential to introduce breaking issues.

If you choose to address these concerns and resubmit, I would be more than happy to merge this.

Once again, thank you for your efforts.

marvin1099 commented:
I have an idea of how to do it actually.
The implementation is going to take some time but here is what i came up with:
If gameprefix_ready and prefix_version not in prefix_wemod_folder
Then symlink gameprefix to wemod_folder
Elif not gameprefix_ready and prefix_version in prefix_wemod_folder
Then copy reqested_version_prefix_in_wemod_folder to gameprefix

To see what version we are working with we allways read the file in the wine prefix that contains the version.

So if we would have lots of sysmlinks in wemod_folder/proton
And if we start a game we check what Versions we have, and copy over the correct on.

A user coice of something like "this proton version is not installied, you want to atempt to use the cosest installed version nr -version_number- and upgrade/downgrade, what may result in some issues"

Since proton will automaticly try to ajust the currend game prefix, so if wemod is installied on a older or newer version the proton autouptate might just work.

What do you think about that, if it sounds good to you i will try to implement something like this.

DaniAsh551 commented:
That sounds good, and not using symlinks makes the task easier (at the cost of some extra storage space).
And `wine` version would not change between minor and revisions of Proton.

Also, on the storage costs topic, we would need some way to remove the prefixes (for proton versions) which are not used by any game. For this, we may need to consult some steam APIs - which would make this tool more difficult to decouple from Steam in the future (which has been requested), but that hardly matters at the moment. We should cross that bridge when we come to it.

marvin1099 commented:
I have revised the implementation to just use the steam compat folder, have a look over it, it should do the job pretty well

marvin1099 commented:
Ok I now have also fixed all but one issue, 
you might have a idea about that issue,
pysimplegui wont show the text I added to it (Line 111+119+127).
Other than that there is basically no more storage usage that before.
This implementation should be good enough for now.
This just uses the prefixes in the steam compat data folder so there is no need for prefix deletion.

marvin1099 commented:
Now after spending lots more time on this,
I added a progress bar to the prefix copy,
Now it only copy's specified folders to make it faster
and so data wont be lost by overwriting important data.

i think now all of your requests, 
on how to make this better,
from 2 days ago are addressed now.

marvin1099 commented:
To adress all your worys and since i had some time:
1. If wemod in not installed anywhere we skip the prefix copy section, this will than continue with your code that will promt the user to Build or Download a prefix
2. If proton versions are the same we are good to go and just copy the prefix (we also copy with progress bar from simplepygui)
3. If there the Versions do not match we ask the user if he wants to continue anyway with the proton version that is cosest, and olny copy on awnser yes (copy progess bar used here too)
4. We can track the version of wineprefixes by just reading the version file of the prefix
5. If the user selects a other version in steam and wemod was installed, nothing has changed, steam will update the prefix and it might work or not
6. If the user selects a other version in steam and wemod is not installied, we scan the steam compat folder and if the same version existsts nr 2. if not nr 3.
7. since we only ever work in the steam compat folder deleting prefixes is not needed since all games have their own prefix anyway and steam manages that
8. We dont ever need to set up a new prefix, steam will manage that and then we just copy the the correct wemod istalled version over (nr 2 and 3.)

I think that is it.
It took me a bit to iron out the bugs, but it seems great now, i think doing it this implementation is a good way of geting this feature.
Go and se what you think.

DaniAsh551 commented:
Hey @marvin1099,

Awesome work! But I still havent had time to sit down and go through the changes, but hopefully I will get some time this Friday.

DaniAsh551 reviewed (approved):
Looks good at the moment
